File: app/src/main/java/jgappsandgames/smartreminderslite/tasks/tags/TagEditorActivity.java
package jgappsandgames.smartreminderslite.tasks.tags;

// JSON
import org.json.JSONArray;
import org.json.JSONException;

// Android OS
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;

// Views
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnLongClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;

// App
import jgappsandgames.smartreminderslite.R;
import jgappsandgames.smartreminderslite.holder.TagHolder.TagSwitcher;
import jgappsandgames.smartreminderslite.utility.ActivityUtility;

// Save
import jgappsandgames.smartreminderssave.tags.TagManager;
import jgappsandgames.smartreminderssave.tasks.Task;

/**
 * Tag Editor Activity
 * Created by joshua on 8/31/17.
 */
public class TagEditorActivity extends Activity implements TextWatcher, OnClickListener, OnLongClickListener, TagSwitcher {
    // Data
    private Task task;

    // Views
    private EditText search_text;
    private Button search_enter;
    private ListView selected;
    private ListView unselected;

    // LifeCycle Methods
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Set Content View
        setContentView(R.layout.activity_tag_editpr);

        // Load Data
        task = new Task(getIntent().getStringExtra(ActivityUtility.TASK_NAME));

        // Find Views
        search_text = findViewById(R.id.search_text);
        search_enter = findViewById(R.id.search_enter);
        selected = findViewById(R.id.selected);
        unselected = findViewById(R.id.unselected);

        // Set Adapters
        selected.setAdapter(new TagSelectedAdapter(this, task));
        unselected.setAdapter(new TagUnselectedAdapter(this, task));

        // Set Listeners
        search_enter.setOnClickListener(this);
        search_enter.setOnLongClickListener(this);
        search_enter.addTextChangedListener(this);

        // Set Result Intent
        setResult(ActivityUtility.RESPONSE_NONE);
    }

    // Text Watcher Methods
    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        // Not Needed, Only included because it is required by the TextWatcher
    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        // Not Needed, OnLy included because it is required by the TextWatcher
    }

    @Override
    public void afterTextChanged(Editable editable) {
        selected.setAdapter(new TagSelectedAdapter(this, task, editable.toString()));
        unselected.setAdapter(new TagUnselectedAdapter(this, task, editable.toString()));

        if (TagManager.contains(editable.toString())) search_enter.setText(R.string.select);
        else search_enter.setText(R.string.add);
    }

    // Click Listeners
    @Override
    public void onClick(View view) {
        if (TagManager.addTag(search_text.getText().toString())) {
            task.addTag(search_text.getText().toString());
            search_text.setText("");

            selected.setAdapter(new TagSelectedAdapter(this, task));
            unselected.setAdapter(new TagUnselectedAdapter(this, task));

            // Set new Result Intent
            try {
                JSONArray tags  = new JSONArray();
                for (String tag : task.getTags()) tags.put(tag);
                setResult(ActivityUtility.RESPONSE_CHANGE, new Intent().putExtra(ActivityUtility.TAG_LIST, tags.toString(4)));
            } catch (JSONException | NullPointerException e) {
                e.printStackTrace();
            }
        } else if (search_text.getText().toString().equals("")) {
            task.addTag(search_text.getText().toString());
            search_text.setText("");

            selected.setAdapter(new TagSelectedAdapter(this, task));
            unselected.setAdapter(new TagUnselectedAdapter(this, task));

            // Set new Result Intent
            try {
                JSONArray tags  = new JSONArray();
                for (String tag : task.getTags()) tags.put(tag);
                setResult(ActivityUtility.RESPONSE_CHANGE, new Intent().putExtra(ActivityUtility.TAG_LIST, tags.toString(4)));
            } catch (JSONException | NullPointerException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public boolean onLongClick(View view) {
        return false;
    }

    // TagSwitcher
    @Override
    public void moveTag(String tag, boolean selected) {
        if (selected) task.addTag(tag);
        else task.removeTag(tag);

        // Set Adapters
        this.selected.setAdapter(new TagSelectedAdapter(this, task));
        this.unselected.setAdapter(new TagUnselectedAdapter(this, task));

        // Set New Result Intent
        try {
            JSONArray tags  = new JSONArray();
            for (String ta : task.getTags()) tags.put(ta);
            setResult(ActivityUtility.RESPONSE_CHANGE, new Intent().putExtra(ActivityUtility.TAG_LIST, tags.toString(4)));
        } catch (JSONException | NullPointerException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/jgappsandgames/smartreminderslite/tasks/tags/TagEditorActivity.kt
package jgappsandgames.smartreminderslite.tasks.tags

// Android
import android.app.Activity
import android.content.Intent
import android.os.Bundle

// Views
import android.text.Editable
import android.text.TextWatcher
import android.view.View

// JSON
import org.json.JSONArray
import org.json.JSONException

// App
import jgappsandgames.smartreminderslite.R
import jgappsandgames.smartreminderslite.adapter.TagAdapter
import jgappsandgames.smartreminderslite.utility.ActivityUtility

// KotlinX
import kotlinx.android.synthetic.main.activity_tag_editpr.tag_editor_search_enter
import kotlinx.android.synthetic.main.activity_tag_editpr.tag_editor_search_text
import kotlinx.android.synthetic.main.activity_tag_editpr.tag_editor_selected
import kotlinx.android.synthetic.main.activity_tag_editpr.tag_editor_unselected

// Save
import jgappsandgames.smartreminderssave.tags.TagManager
import jgappsandgames.smartreminderssave.tasks.Task

/**
 * TagEditorActivity
 * Created by joshua on 1/19/2018.
 */
class TagEditorActivity: Activity(), TextWatcher, View.OnClickListener, View.OnLongClickListener, TagAdapter.TagSwitcher {
    // Data ----------------------------------------------------------------------------------------
    private lateinit var task: Task

    // LifeCycle Methods --------------------------------------------------------------------------
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tag_editpr)

        // Set Result Intent
        setResult(ActivityUtility.RESPONSE_NONE)

        // Load Data
        task = Task(intent.getStringExtra(ActivityUtility.TASK_NAME))

        // Set Listeners
        tag_editor_search_enter.setOnClickListener(this)
        tag_editor_search_enter.setOnLongClickListener(this)
        tag_editor_search_text.addTextChangedListener(this)

        // Set Adapters
        tag_editor_selected.adapter = TagAdapter(this, this, task.getTags(), true)
        tag_editor_unselected.adapter = TagAdapter(this, this, task.getTags(), false)
    }

    // Text Watcher Methods ------------------------------------------------------------------------
    override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
    override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
        if (tag_editor_search_text.text.toString().isEmpty()) {
            tag_editor_selected!!.adapter = TagAdapter(this, this, task.getTags(), true)
            tag_editor_unselected!!.adapter = TagAdapter(this, this, task.getTags(), false)
        } else {
            tag_editor_selected!!.adapter = TagAdapter(this, this, task.getTags(), true, tag_editor_search_text.text.toString())
            tag_editor_unselected!!.adapter = TagAdapter(this, this, task.getTags(), false, tag_editor_search_text.text.toString())
        }

        if (TagManager.contains(tag_editor_search_text.text.toString())) tag_editor_search_enter.setText(R.string.select)
        else tag_editor_search_enter.setText(R.string.add)
    }
    override fun afterTextChanged(editable: Editable) {}

    // Click Listeners -----------------------------------------------------------------------------
    override fun onClick(view: View) {
        // Tag is Not in the List And is addable
        if (TagManager.addTag(tag_editor_search_text.text.toString())) {
            task.addTag(tag_editor_search_text.text.toString())
            tag_editor_search_text.setText("")

            tag_editor_selected!!.adapter = TagAdapter(this, this, task.getTags(), true)
            tag_editor_unselected!!.adapter = TagAdapter(this, this, task.getTags(), false)

            // Set new Result Intent
            try {
                val tags = JSONArray()
                for (tag in task.getTags()) tags.put(tag)
                setResult(ActivityUtility.RESPONSE_CHANGE, Intent().putExtra(ActivityUtility.TAG_LIST, tags.toString()))
            } catch (e: JSONException) {
                e.printStackTrace()
            } catch (e: NullPointerException) {
                e.printStackTrace()
            }
        } else if (tag_editor_search_text.text.toString() != "") {
            task.addTag(tag_editor_search_text.text.toString())
            tag_editor_search_text.setText("")

            tag_editor_selected!!.adapter = TagAdapter(this, this, task.getTags(), true)
            tag_editor_unselected!!.adapter = TagAdapter(this, this, task.getTags(), false)

            // Set new Result Intent
            try {
                val tags = JSONArray()
                for (tag in task.getTags()) tags.put(tag)
                setResult(ActivityUtility.RESPONSE_CHANGE, Intent().putExtra(ActivityUtility.TAG_LIST, tags.toString()))
            } catch (e: JSONException) {
                e.printStackTrace()
            } catch (e: NullPointerException) {
                e.printStackTrace()
            }
        }
    }

    override fun onLongClick(view: View): Boolean {
        return false
    }

    // TagSwitcher ---------------------------------------------------------------------------------
    override fun moveTag(tag: String?, selected: Boolean) {
        if (selected) task.addTag(tag!!)
        else task.removeTag(tag!!)

        // Set Adapters
        this.tag_editor_selected!!.adapter = TagAdapter(this, this, task.getTags(), true)
        this.tag_editor_unselected!!.adapter = TagAdapter(this, this, task.getTags(), false)

        // Set New Result Intent
        try {
            val tags = JSONArray()
            for (ta in task.getTags()) tags.put(ta)
            setResult(ActivityUtility.RESPONSE_CHANGE, Intent().putExtra(ActivityUtility.TAG_LIST, tags.toString()))
        } catch (e: JSONException) {
            e.printStackTrace()
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
    }
}